### PR TITLE
Update CLIP to a functional model

### PR DESCRIPTION
**Pull request by divyashreepathihalli**

### CODE REVIEW

File: keras_cv/models/feature_extractor/clip/clip_model.py [hunk 1]
```diff
@@ -180,8 +157,52 @@ def call(self, inputs):
...
+        self.transformer_heads = transformer_heads
+        self.transformer_layers = transformer_layers
 
-        return logits_per_image, logits_per_text
+    def build(self, input_shape):
+        super().build(input_shape)
+        self.text_encoder.build([None, self.context_length])
+        self.image_encoder.build(
+            [None, self.image_resolution, self.image_resolution, 3]
+        )
```
Comment (tirthasheshpatel) [inline keras_cv/models/feature_extractor/clip/clip_model.py:190]: I think you can remove the build and compute output shape method now. Only layers/models that are not implemented in the functional manner need them.

File: keras_cv/models/feature_extractor/clip/clip_model.py [hunk 2]
```diff
@@ -103,65 +103,42 @@ def __init__(
...
+        vision_heads = vision_width // 64
+        image_input = keras.layers.Input(
+            shape=(None, image_resolution, image_resolution, 3), name="image"
+        )
+        text_input = keras.layers.Input(
+            shape=(None, context_length), name="text"
+        )
+        attention_mask_input = keras.layers.Input(
+            shape=(None, context_length), name="attention_mask"
+        )
```
Comment (tirthasheshpatel) [inline keras_cv/models/feature_extractor/clip/clip_model.py:116]:
```suggestion
        image_input = keras.layers.Input(
            shape=(image_resolution, image_resolution, 3), name="image"
        )
        text_input = keras.layers.Input(
            shape=(context_length,), name="text"
        )
        attention_mask_input = keras.layers.Input(
            shape=(context_length,), name="attention_mask"
        )
```

**Comment by tirthasheshpatel** [comment]
Is our test suite running with Tf 1.16? I keep getting this error constantly when testing with TensorFlow 1.16.1:

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[15], line 1
----> 1 processor = CLIPProcessor(
      2     MODEL_CONFIGS[config_name]["image_resolution"], VOCAB_PATH, MERGE_PATH
      3 )
      4 image = processor.process_images(["two_cats.jpg"])
      5 text_input = ["mountains", "cat on tortoise", "two cats"]

File ~/oss/keras-cv/keras_cv/models/feature_extractor/clip/clip_processor.py:58, in CLIPProcessor.__init__(self, input_resolution, vocabulary, merges, **kwargs)
     51 self.image_transform = self.transform_image
     52 self.tokenizer = CLIPTokenizer(
     53     vocabulary=self.vocabulary,
     54     merges=self.merges,
     55     unsplittable_tokens=["</w>"],
     56 )
     57 self.packer = StartEndPacker(
---> 58     start_value=self.tokenizer.token_to_id("<|startoftext|>"),
     59     end_value=self.tokenizer.token_to_id("<|endoftext|>"),
     60     pad_value=None,
     61     sequence_length=77,
     62     return_padding_mask=True,
     63 )

File ~/oss/virtualenvs/keras-cv-dev/lib/python3.11/site-packages/keras_nlp/src/tokenizers/byte_pair_tokenizer.py:420, in BytePairTokenizer.token_to_id(self, token)
    418 """Convert a string token to an integer id."""
    419 self._check_vocabulary()
--> 420 return self.vocabulary[token]

KeyError: '<|startoftext|>'
```

**Comment by divyashreepathihalli** [comment]
> Is our test suite running with Tf 1.16? I keep getting this error constantly when testing with TensorFlow 1.16.1:
> 
> ```
> ---------------------------------------------------------------------------
> KeyError                                  Traceback (most recent call last)
> Cell In[15], line 1
> ----> 1 processor = CLIPProcessor(
>       2     MODEL_CONFIGS[config_name]["image_resolution"], VOCAB_PATH, MERGE_PATH
>       3 )
>       4 image = processor.process_images(["two_cats.jpg"])
>       5 text_input = ["mountains", "cat on tortoise", "two cats"]
> 
> File ~/oss/keras-cv/keras_cv/models/feature_extractor/clip/clip_processor.py:58, in CLIPProcessor.__init__(self, input_resolution, vocabulary, merges, **kwargs)
>      51 self.image_transform = self.transform_image
>      52 self.tokenizer = CLIPTokenizer(
>      53     vocabulary=self.vocabulary,
>      54     merges=self.merges,
>      55     unsplittable_tokens=["</w>"],
>      56 )
>      57 self.packer = StartEndPacker(
> ---> 58     start_value=self.tokenizer.token_to_id("<|startoftext|>"),
>      59     end_value=self.tokenizer.token_to_id("<|endoftext|>"),
>      60     pad_value=None,
>      61     sequence_length=77,
>      62     return_padding_mask=True,
>      63 )
> 
> File ~/oss/virtualenvs/keras-cv-dev/lib/python3.11/site-packages/keras_nlp/src/tokenizers/byte_pair_tokenizer.py:420, in BytePairTokenizer.token_to_id(self, token)
>     418 """Convert a string token to an integer id."""
>     419 self._check_vocabulary()
> --> 420 return self.vocabulary[token]
> 
> KeyError: '<|startoftext|>'
> ```

This needs to be a conditional import, let me send a fix

**Comment by divyashreepathihalli** [comment]
https://github.com/keras-team/keras-cv/pull/2396

**Comment by tirthasheshpatel** [comment]
In the latest commit, I have removed some numerics tests since we don't match the HF model. Also, updated the presets and squashed everything down to one commit. Sorry, I lost authorship when doing this. I will approve but leave merging up to @divyashreepathihalli since I have a lot of contributions to this. Please let me know if you have any changes in mind, otherwise feel free to merge. Thanks!

File: keras_cv/models/feature_extractor/clip/clip_model_test.py [hunk 3]
```diff
@@ -42,98 +42,77 @@ def test_clip_model_golden_values(self):
...
+            outputs["text_logits"],
+            ops.transpose([[10.246354, 10.246353, 10.246354]]),
         )
 
+        # True reference values computed using HF:
+        # image_logits: [[17.8013, 17.8013, 17.8013]]
+        # text_logits: image_logits.T
+
+        # xfail after assertion
+        pytest.xfail("KerasCV CLIP doesn't match the HF model.")
```
Comment (tirthasheshpatel) [inline keras_cv/models/feature_extractor/clip/clip_model_test.py:72]: @divyashreepathihalli I have included the reference values here computed using the same inputs to the HF model. Currently, `xfail`ing the test to indicate we need to fix this.

**Review by tirthasheshpatel — APPROVED** [review]
I verified the refactor has the same numerics as the model we had before. But note that those numerics are broken too, we are very far from the equivalent HF model.

**Comment by tirthasheshpatel** [comment]
Unrelated Keras 2 failure. @divyashreepathihalli Do we need to the initializer? I am inclined to remove it since we don't yet support training.

**Comment by divyashreepathihalli** [comment]
> Unrelated Keras 2 failure. @divyashreepathihalli Do we need to the initializer? I am inclined to remove it since we don't yet support training.

yes! we can remove those.

**Pull request merged**